src/spine41.js

```diff
@@ -0,0 +1,669 @@
+            var batchStartIndex = 0;
+            var batchIndexCount = 0;
+
+            // vertex / index buffer access
+            var dstVertices = new pc.VertexIterator(this._vertexBuffer);
+            var dstIndices = new Uint16Array(this._indexBuffer.lock());
+            var dstIndexOffset = 0, dstVertexOffset = 0;
+
+            var drawOrder = this.skeleton.drawOrder;
+            var count = drawOrder.length;
+            for (var i = 0; i < count; i++) {
+                var slot = drawOrder[i];
+
+                if (slot.attachment && slot.material && slot.positions.length > 0 && slot.indices.length > 0) {
+
+                    // material switch
+                    if (currentMaterialKey && currentMaterialKey !== slot.material) {
+                        this.SubmitBatch(batchStartIndex, batchIndexCount, currentMaterialKey);
+
+                        // prepare next batch
+                        currentMaterialKey = slot.material;
+                        batchStartIndex = dstIndexOffset;
+                        batchIndexCount = 0;
+
+                    }
+                    currentMaterialKey = slot.material;
+
+                    // write vertex data
+                    var positions = slot.positions;
+                    var r = slot.vertexColor.r, g = slot.vertexColor.g, b = slot.vertexColor.b, a = slot.vertexColor.a;
+                    var uvs = slot.uvs;
+                    var j, posCount = positions.length / 2;
+
+                    for (j = 0; j < posCount; j++) {
+                        dstVertices.element[pc.SEMANTIC_POSITION].set(positions[j * 2], positions[j * 2 + 1]);
+                        dstVertices.element[pc.SEMANTIC_NORMAL].set(0, 255, 0, 0);      // 0,1,0 normal stored in 8 bit per component
+                        dstVertices.element[pc.SEMANTIC_COLOR].set(r, g, b, a);
+                        dstVertices.element[pc.SEMANTIC_TEXCOORD0].set(uvs[j * 2], 1.0 - uvs[j * 2 + 1]);
+                        dstVertices.next();
+                    }
+
+                    // write index data
+                    var indices = slot.indices;
+                    var indCount = indices.length;
+                    for (j = 0; j < indCount; j++)
+                        dstIndices[dstIndexOffset + j] = indices[j] + dstVertexOffset;
+
+                    batchIndexCount += indCount;
+                    dstIndexOffset += indCount;
+                    dstVertexOffset += posCount;
+                }
+            }
+
+            dstVertices.end();
+            this._indexBuffer.unlock();
+
+            // final batch
+            this.SubmitBatch(batchStartIndex, batchIndexCount, currentMaterialKey);
+        }
+
+        // add all instances to layers
+        this.addToLayers();
+    };
+
+    Spine.prototype.SubmitBatch = function (indexBase, indexCount, materialKey) {
+        if (indexCount > 0) {
+            var mesh = new pc.Mesh(this._app.graphicsDevice);
```

Comment:
You seem to be passing the graphics device to the `Mesh` constructor now, but the other copy of spine.js doesn't have this fix. Might even be worth having this in a separate PR since it's unrelated.

Reply:
I need that so this builds and runs without warnings on 1.57 and above.

Reply:
The `Mesh` constructor switched to accepting a `GraphicsDevice` as a parameter over 3 years ago:

https://github.com/playcanvas/engine/pull/1960/files#diff-f68894866f45f9fc28805640c3670b3865dc50dc44037b6d6fb602fe86726924R144

All `Mesh` constructors across the source should be updated based on the current `Mesh` constructor signature. I don't see how anything could break.

Reply:
You right !  The last commit was about line endings not changing the signature.  Yeah, lets bring it in when we refactor.  #70